### ISSUE TITLE
Nexus pod crashing and GC Previews erroring

### ISSUE THREAD

**Issue by polothy** [jenkins-x/jx#3451]
### Summary 

Upgraded platform and now `nexus` pod is in `CrashLoopBackOff` state.  In addition, `jenkins-x-gcpreviews` pods are in `Error` state.

### Steps to reproduce the behavior

Just upgraded platform: `jx upgrade platform --verbose --update-secrets --log-level=debug`

Did these two upgrade steps:
* Upgrading platform from version 0.0.3515 to version 0.0.3636
* Upgrading platform from version 0.0.3636 to version 0.0.3657

### Expected behavior

Pods bring joy to our lives!

#### Nexus pod

The `nexus` pod crash seems to be failing due to the post hook: `PostStartHookError: command '/opt/sonatype/nexus/postStart.sh' exited with 1: ERROR
: Login to nexus failed. Tried both the default password and the provided password secret file.`

I'm not entirely sure if the error is because of an invalid password or because it cannot reach the Nexus server to actually login.

Here are some pod logs:
```
2019-03-27 21:01:55,589+0000 ERROR [quartz-3-thread-15] *SYSTEM org.quartz.core.ErrorLogger - Unable to notify JobListener(s) of Job that was executed: (error will be ignored). trigger= nexus.b51451e5-a1ed-4e93-8228-23f28d210fa4 job= nexus.b51451e5-a1ed-4e93-8228-23f28d210fa4
org.quartz.SchedulerException: JobListener 'org.sonatype.nexus.quartz.internal.task.QuartzTaskJobListener:nexus.b51451e5-a1ed-4e93-8228-23f28d210fa4' threw exception: Invalid state: STOPPED; allowed: [STARTED]
	at org.quartz.core.QuartzScheduler.notifyJobListenersWasExecuted(QuartzScheduler.java:1963)
	at org.quartz.core.JobRunShell.notifyJobListenersComplete(JobRunShell.java:340)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:224)
	at org.sonatype.nexus.quartz.internal.QuartzThreadPool.lambda$0(QuartzThreadPool.java:143)
	at org.sonatype.nexus.thread.internal.MDCAwareRunnable.run(MDCAwareRunnable.java:40)
	at org.apache.shiro.subject.support.SubjectRunnable.doRun(SubjectRunnable.java:120)
	at org.apache.shiro.subject.support.SubjectRunnable.run(SubjectRunnable.java:108)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.sonatype.nexus.common.stateguard.InvalidStateException: Invalid state: STOPPED; allowed: [STARTED]
	at org.sonatype.nexus.common.stateguard.StateGuard._ensure(StateGuard.java:115)
	at org.sonatype.nexus.common.stateguard.StateGuard.access$1(StateGuard.java:108)
	at org.sonatype.nexus.common.stateguard.StateGuard$GuardImpl.run(StateGuard.java:271)
	at org.sonatype.nexus.common.stateguard.GuardedInterceptor.invoke(GuardedInterceptor.java:53)
	at org.sonatype.nexus.quartz.internal.task.QuartzTaskInfo.setNexusTaskState(QuartzTaskInfo.java:134)
	at org.sonatype.nexus.quartz.internal.task.QuartzTaskJobListener.jobWasExecuted(QuartzTaskJobListener.java:211)
	at org.quartz.core.QuartzScheduler.notifyJobListenersWasExecuted(QuartzScheduler.java:1961)
	... 11 common frames omitted
2019-03-27 21:01:55,590+0000 INFO  [quartz-3-thread-13] *SYSTEM org.sonatype.nexus.quartz.internal.task.QuartzTaskInfo - Task 'Repair - Rebuild repository search' [repository.rebuild-index] state change RUNNING -> DONE (FAILED)
2019-03-27 21:01:55,591+0000 ERROR [quartz-3-thread-13] *SYSTEM org.quartz.core.ErrorLogger - Unable to notify JobListener(s) of Job that was executed: (error will be ignored). trigger= nexus.af8cacc5-4fcb-4f19-9b3d-8879b8fbd777 job= nexus.af8cacc5-4fcb-4f19-9b3d-8879b8fbd777
org.quartz.SchedulerException: JobListener 'org.sonatype.nexus.quartz.internal.task.QuartzTaskJobListener:nexus.af8cacc5-4fcb-4f19-9b3d-8879b8fbd777' threw exception: Invalid state: STOPPED; allowed: [STARTED]
	at org.quartz.core.QuartzScheduler.notifyJobListenersWasExecuted(QuartzScheduler.java:1963)
	at org.quartz.core.JobRunShell.notifyJobListenersComplete(JobRunShell.java:340)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:224)
	at org.sonatype.nexus.quartz.internal.QuartzThreadPool.lambda$0(QuartzThreadPool.java:143)
	at org.sonatype.nexus.thread.internal.MDCAwareRunnable.run(MDCAwareRunnable.java:40)
	at org.apache.shiro.subject.support.SubjectRunnable.doRun(SubjectRunnable.java:120)
	at org.apache.shiro.subject.support.SubjectRunnable.run(SubjectRunnable.java:108)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.sonatype.nexus.common.stateguard.InvalidStateException: Invalid state: STOPPED; allowed: [STARTED]
	at org.sonatype.nexus.common.stateguard.StateGuard._ensure(StateGuard.java:115)
	at org.sonatype.nexus.common.stateguard.StateGuard.access$1(StateGuard.java:108)
	at org.sonatype.nexus.common.stateguard.StateGuard$GuardImpl.run(StateGuard.java:271)
	at org.sonatype.nexus.common.stateguard.GuardedInterceptor.invoke(GuardedInterceptor.java:53)
	at org.sonatype.nexus.quartz.internal.task.QuartzTaskInfo.setNexusTaskState(QuartzTaskInfo.java:134)
	at org.sonatype.nexus.quartz.internal.task.QuartzTaskJobListener.jobWasExecuted(QuartzTaskJobListener.java:211)
	at org.quartz.core.QuartzScheduler.notifyJobListenersWasExecuted(QuartzScheduler.java:1961)
	... 11 common frames omitted
2019-03-27 21:01:57,205+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.extender.NexusLifecycleManager - Stop SECURITY
2019-03-27 21:01:57,207+0000 INFO  [FelixStartLevel] *SYSTEM org.apache.shiro.session.mgt.AbstractValidatingSessionManager - Disabled session validation scheduler.
2019-03-27 21:01:57,207+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.extender.NexusLifecycleManager - Stop EVENTS
2019-03-27 21:01:57,207+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.extender.NexusLifecycleManager - Stop SCHEMAS
2019-03-27 21:01:57,213+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.extender.NexusLifecycleManager - Stop UPGRADE
2019-03-27 21:01:57,213+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.extender.NexusLifecycleManager - Stop RESTORE
2019-03-27 21:01:57,213+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.extender.NexusLifecycleManager - Stop STORAGE
2019-03-27 21:01:57,214+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping 5 pools
2019-03-27 21:01:57,214+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping pool: security
2019-03-27 21:01:57,216+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping pool: component
2019-03-27 21:01:57,217+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping pool: audit
2019-03-27 21:01:57,217+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping pool: config
2019-03-27 21:01:57,217+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping pool: accesslog
2019-03-27 21:01:57,217+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping 5 instances
2019-03-27 21:01:57,217+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping instance: security
2019-03-27 21:01:57,217+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping instance: component
2019-03-27 21:01:57,217+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping instance: audit
2019-03-27 21:01:57,217+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping instance: config
2019-03-27 21:01:57,218+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseManagerImpl - Stopping instance: accesslog
2019-03-27 21:01:57,218+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.server.OServer - OrientDB Server is shutting down...
2019-03-27 21:01:57,218+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.server.OServer - Shutting down protocols
2019-03-27 21:01:57,219+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.server.plugin.OServerPluginManager - Shutting down plugins:
2019-03-27 21:01:57,220+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.server.plugin.OServerPluginManager - - jmx
2019-03-27 21:01:57,220+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.server.OServer - Shutting down databases:
2019-03-27 21:01:57,221+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.core.Orient - Orient Engine is shutting down...
2019-03-27 21:01:57,224+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.core.Orient - - shutdown storage: security...
2019-03-27 21:01:57,263+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.core.Orient - - shutdown storage: component...
2019-03-27 21:01:57,290+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.core.Orient - - shutdown storage: audit...
2019-03-27 21:01:57,326+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.core.Orient - - shutdown storage: config...
2019-03-27 21:01:57,359+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.core.Orient - - shutdown storage: OSystem...
2019-03-27 21:01:57,359+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.core.Orient - - shutdown storage: accesslog...
2019-03-27 21:01:57,916+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.core.Orient - OrientDB Engine shutdown complete
2019-03-27 21:01:57,917+0000 INFO  [FelixStartLevel] *SYSTEM com.orientechnologies.orient.server.OServer - OrientDB Server shutdown complete
2019-03-27 21:01:57,922+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.internal.orient.DatabaseServerImpl - Shutdown
2019-03-27 21:01:57,924+0000 INFO  [FelixStartLevel] *SYSTEM org.elasticsearch.node - [91D230B4-DF05C580-57F384EF-4DA22AA6-E1476F7B] stopping ...
2019-03-27 21:01:57,973+0000 INFO  [FelixStartLevel] *SYSTEM org.elasticsearch.node - [91D230B4-DF05C580-57F384EF-4DA22AA6-E1476F7B] stopped
2019-03-27 21:01:57,973+0000 INFO  [FelixStartLevel] *SYSTEM org.elasticsearch.node - [91D230B4-DF05C580-57F384EF-4DA22AA6-E1476F7B] closing ...
2019-03-27 21:01:57,979+0000 INFO  [FelixStartLevel] *SYSTEM org.elasticsearch.node - [91D230B4-DF05C580-57F384EF-4DA22AA6-E1476F7B] closed
2019-03-27 21:01:57,984+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'NexusAuthorizingRealm.authorizationCache' removed from EhcacheManager.
2019-03-27 21:01:57,985+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'nuget.org-proxy#odata-query-cache' removed from EhcacheManager.
2019-03-27 21:01:57,985+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'jenkins-release#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,986+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'nuget.org-proxy#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,986+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'spring-release#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,987+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'assetDownloadCountStoreCache' removed from EhcacheManager.
2019-03-27 21:01:57,987+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'jitpack#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,988+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'maven-jenkinsci#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,988+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'nuget-group#odata-query-cache' removed from EhcacheManager.
2019-03-27 21:01:57,988+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'npmjs-registry#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,989+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'jenkins-public#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,989+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'spring-milestone#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,990+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'NexusAuthenticatingRealm.authenticationCache' removed from EhcacheManager.
2019-03-27 21:01:57,990+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'maven-central#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,990+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'gradle-distributions#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,991+0000 INFO  [FelixStartLevel] *SYSTEM org.ehcache.core.EhcacheManager - Cache 'plugins-gradle#negative-cache' removed from EhcacheManager.
2019-03-27 21:01:57,992+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.cache.internal.ehcache.EhCacheManagerProvider - Cache-manager closed
2019-03-27 21:01:57,993+0000 INFO  [FelixStartLevel] *SYSTEM org.sonatype.nexus.extender.NexusLifecycleManager - Stop KERNEL
```

#### jenkins-x-gcpreviews pods

The `jenkins-x-gcpreviews` pods are in `Error` state.  This one is more straightforward, pod logs show:

```
error: namespaces "jxdevptfm" is forbidden: User "system:serviceaccount:jxdevptfm:jenkins-x-gcpreviews" cannot get namespaces in the namespace "jxdevptfm"
```

### Jx version

The output of `jx version` is:

```
NAME               VERSION
jx                 1.3.1032
jenkins x platform 0.0.3657
Kubernetes cluster v1.11.7
kubectl            v1.13.4
helm client        v2.12.2+g7d2b0c7
helm server        v2.12.2+g7d2b0c7
git                git version 2.21.0
Operating System   Mac OS X 10.14.3 build 18D109
```

### Jenkins type

- [x] Classic Jenkins
- [ ] Serverless Jenkins

### Kubernetes cluster

Kops cluster in AWS.  We create the cluster via kops instead of going through JX.

### Operating system / Environment

macOS

**Comment by polothy** [comment]
Just wanted to add, the `nexus` secret has the correct password in it.

**Comment by warrenbailey** [comment]
I can't seem to replicate the nexus issue on my cluster, but I have the same problem with the jenkins-x-gcpreviews pods.

**Comment by polothy** [comment]
Any tips on nexus would be appreciated!  Not sure how to even debug it or anything to try.

**Comment by sboardwell** [comment]
If you delete the Nexus pod you see the "startup" logs. Which version of nexus is being used? There have been improvements to the `postStart.sh` in the past months.

**Comment by polothy** [comment]
Posted full logs here: https://gist.github.com/polothy/ac9428879f6cf7e5eb895f081b49d066

If you look at the [2nd file](https://gist.githubusercontent.com/polothy/ac9428879f6cf7e5eb895f081b49d066/raw/0e7da225c917426f67a97b3a49efbe9749ecd7de/rest-of-log) in the gist, where it is starting `Started Sonatype Nexus OSS 3.15.1-01`, it looks like it is failing to rebuild an index:

```
Failed to run task 'Rebuilding search index of *' on repository 'plugins-gradle'
```

Hopefully this gives someone a clue on how to go about resolving or some sort of workaround.

**Comment by sboardwell** [comment]
I think the `2019-03-28 13:16:13,968+0000 WARN  [SIGTERM handler] *SYSTEM com.orientechnologies.orient.core.OSignalHandler - Received signal: SIGTERM` a few lines before might be the cause of that.

**Comment by polothy** [comment]
So, @sboardwell has been helping me in Slack.  First, we thought maybe Nexus was failing due to resource limits.  We increased JVM memory limit to 4GB and increased limits on the deployment yaml:
```yaml
        resources:
          limits:
            cpu: "1"
            memory: 4Gi
          requests:
            cpu: 500m
            memory: 2Gi
```

**The above did not work, BUT**

@sboardwell suggested this change to the deployment yaml:
```
lifecycle:
          postStart:
            exec:
              command:
              - /opt/sonatype/nexus/postStart.sh
```
to
```
lifecycle:
          postStart:
            exec:
              command:
              - echo
```
The nexus pod starts running!  So, then I ssh'ed into the pod and tried some of the curls in [postStart.sh](https://github.com/jenkins-x-charts/nexus/blob/master/postStart.sh#L61) and the default admin password and our password both don't work.  Get a `HTTP/1.1 401 Unauthorized` response.  Also, our password in `cat /opt/sonatype/nexus/config/password` is correct.

So, the question is, how does Nexus get its credentials?

**Comment by polothy** [comment]
Resolved Nexus issue: with the above hack, the pod is running.  The URL for Nexus was working, so I went there and started trying to login.  I found a _very_ old password and was able to login.  Then I changed the password to our current password.  Now regular Nexus deployment worked!

I'm not certain what happened, but I do recall that this particular install had the password change.  So, Nexus held onto the old password somehow.
